selfdrive/boardd/tests/test_boardd_usbprotocol.cc

```diff
@@ -0,0 +1,127 @@
+#define CATCH_CONFIG_MAIN
+#define CATCH_CONFIG_ENABLE_BENCHMARKING
+#include <random>
+
+#include "catch2/catch.hpp"
+#include "cereal/messaging/messaging.h"
+#include "selfdrive/boardd/panda.h"
+
+const int CHUNK_SIZE = 64;
+const unsigned char dlc_to_len[] = {0U, 1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 12U, 16U, 20U, 24U, 32U, 48U, 64U};
+
+int random_int(int min, int max) {
+  std::random_device dev;
+  std::mt19937 rng(dev());
+  std::uniform_int_distribution<std::mt19937::result_type> dist(min, max);
+  return dist(rng);
+}
+
+struct PandaTest : public Panda {
+  PandaTest(uint32_t bus_offset, int can_list_size);
+  void test_can_send();
+  void test_can_recv();
+
+  std::map<int, std::string> test_data;
+  int can_list_size = 0;
+  int total_pakets_size = 0;
+  MessageBuilder msg;
+  capnp::List<cereal::CanData>::Reader can_data_list;
+};
+
+PandaTest::PandaTest(uint32_t buf_offset_, int can_list_size) : can_list_size(can_list_size), Panda(buf_offset_) {
```

Comment:
buf_offset ? typo?

Reply:
ah,yes.-_-

Reply:
Also it should have underscore at the end? Strange why it didn't failed with typo if you use bus_offset in setting .src

Reply:
because we send it to base class's CTOR, we have to use a different name with the class member "bus_offset",   otherwise a compile error will be thrown